docs/spec_draft.md

```diff
@@ -1808,6 +1809,68 @@ where `i[d] = j[permutation[d]]`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.while
+
+### Semantics
+
+Produces the output from executing `body` function 0 or more times while the
+`cond` function outputs `true`. More formally, the semantics can be expressed
+using Python-like syntax as follows:
+
+```python
+internal_state = operands
+while cond(internal_state) == True:
+  internal_state = body(internal_state)
+results = internal_state
+```
+
+The behaviour of an infinite loop is implementation-defined.
```

Comment:
I agree with Mehdi that "undefined behavior" and "implementation-defined behavior" mean two different things, and undefined behavior is not a good fit here.

From https://en.wikipedia.org/wiki/Undefined_behavior: "In computer programming, undefined behavior (UB) is the result of executing a program whose behavior is prescribed to be unpredictable, in the language specification to which the computer code adheres. This is different from unspecified behavior, for which the language specification does not prescribe a result, and implementation-defined behavior that defers to the documentation of another component of the platform (such as the ABI or the translator documentation)".

"The behaviour of an infinite loop is implementation-defined" seems like a good fit for the spec draft, given that it: 1) calls out a corner case, 2) specifies its behavior in a way that's consistent with existing implementations without precluding further refinements. I think it would be okay to merge this PR and leave the rest to #383, but let's wait for Mehdi's input before doing that.

@joker-eph Do you have objections to the current wording? Or maybe advice on how to define this better for this op and/or at the scope of the entire spec?

Reply:
I don't understand why you think that UB isn't a good fit here?

Reply:
As you said in the nearby comment thread, termination requires careful consideration. Saying that infinite loops are UB is one way to provide a definitive ruling on this, but I think there are other reasonable alternatives if we want to go deeper into things like memory / thread model.

Correct me if I'm wrong, but in my mind "implementation-defined" is the most non-committal way to describe this behavior if we don't want to definitively decide on this at the moment, hence the proposed wording.

Reply:
Actually no: undefined behavior is the least committal way of describe anything: it basically forbids this in the input. It's like saying "it isn't permitted".
"Implementation-defined" means that StableHLO supports it and no generic code can assume it won't be present.

Reply:
Looks like we might be using different words to describe similar things. What I meant is that calling this implementation-defined seems to be the least amount of guarantees that we can make to consumers at the time. 

If in the future, as part of working on #383 and #384, we'll want to turn non-termination into UB (and I think we likely will), then we'll be strengthening these guarantees. This seems easier to do than go the other way around - call this UB right now, and if upon closer inspection we decide that non-termination is not UB, then relax the guarantees.

Reply:
> Looks like we might be using different words to describe similar things.

Not clear to me: we can't be using "implementation defined" and "undefined behavior" for the same thing...

> What I meant is that calling this implementation-defined seems to be the least amount of guarantees that we can make to consumers at the time.

I am not sure I understand this sentence, maybe because "consumers" isn't clear to me here?

> This seems easier to do than go the other way around - call this UB right now, and if upon closer inspection we decide that non-termination is not UB, then relax the guarantees.

I have the complete opposite view on this. You can't make something UB later, but you can define it later.

Reply:
"we can't be using "implementation defined" and "undefined behavior" for the same thing". I think you and I both mean the same thing for "implementation defined" and "undefined behavior", but we might have a difference in opinions on what "the least committal" means in this case. Let's discuss this in a higher-bandwidth setting and follow up with a resolution here afterwards.

Reply:
Please change to "The behaviour of an infinite loop is TBD" as discussed above. After that, we can merge.